integration-cli/docker_cli_build_test.go

```diff
@@ -5438,3 +5438,282 @@ func (s *DockerSuite) TestBuildEmptyStringVolume(c *check.C) {
 	}
 
 }
+
+func (s *DockerSuite) TestBuildBuildTimeEnv(c *check.C) {
+	imgName := "bldenvtest"
+	defer func() { deleteImages(imgName) }()
+	envKey := "foo"
+	envVal := "bar"
+	buildCmd := exec.Command(dockerBinary, "build", "-t", imgName, "--build-env",
+		fmt.Sprintf("%s=%s", envKey, envVal), "-")
+	buildCmd.Stdin = strings.NewReader(fmt.Sprintf("FROM busybox\n"+
+		"RUN echo $%s\n"+
+		"CMD echo $%s\n",
+		envKey, envKey))
+
+	if out, _, err := runCommandWithOutput(buildCmd); err != nil || !strings.Contains(out, envVal) {
+		if err != nil {
+			c.Fatalf("build failed to complete: %v %v", out, err)
+		}
+		c.Fatalf("failed to access environment variable in output: '%v' "+
+			"expected: '%v'", out, envVal)
+	}
+
+	containerName := "bldenvCont"
+	defer func() { deleteContainer(containerName) }()
+	runCmd := exec.Command(dockerBinary, "run", "--name", containerName, imgName)
+	if out, _, err := runCommandWithOutput(runCmd); out != "\n" || err != nil {
+		c.Fatalf("run produced invalid output: %q, expected %q", out, "")
+	}
+}
+
+func (s *DockerSuite) TestBuildBuildTimeEnvCacheHit(c *check.C) {
+	imgName := "bldenvtest"
+	defer func() { deleteImages(imgName) }()
+	envKey := "foo"
+	envVal := "bar"
+
+	ctx, err := fakeContext(fmt.Sprintf("FROM busybox\n"+
+		"RUN echo $%s\n", envKey), map[string]string{})
+	defer ctx.Close()
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	args := []string{
+		"--build-env", fmt.Sprintf("%s=%s", envKey, envVal),
+	}
+	if _, err := buildImageFromContextWithArgs(imgName, ctx, args...); err != nil {
+		c.Fatal(err)
+	}
+
+	imgNameCache := "bldenvtestcachehit"
+	usingCacheStr := "Using cache"
+	defer func() { deleteImages(imgNameCache) }()
+	buildCmd := exec.Command(dockerBinary, "build", "-t", imgNameCache, "--build-env",
+		fmt.Sprintf("%s=%s", envKey, envVal), "-")
+	buildCmd.Stdin = strings.NewReader(fmt.Sprintf("FROM busybox\n"+
+		"RUN echo $%s\n", envKey))
+
+	if out, _, err := runCommandWithOutput(buildCmd); err != nil || !strings.Contains(out, usingCacheStr) {
+		if err != nil {
+			c.Fatalf("build failed to complete: %v %v", out, err)
+		}
+		c.Fatalf("build didn't use cache! output: '%v' expected to contain: '%v'", out, usingCacheStr)
+	}
+}
+
+func (s *DockerSuite) TestBuildBuildTimeEnvCacheMiss(c *check.C) {
+	imgName := "bldenvtest"
+	defer func() { deleteImages(imgName) }()
+	envKey := "foo"
+	envVal := "bar"
+
+	ctx, err := fakeContext(fmt.Sprintf("FROM busybox\n"+
+		"RUN echo $%s\n", envKey), map[string]string{})
+	defer ctx.Close()
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	args := []string{
+		"--build-env", fmt.Sprintf("%s=%s", envKey, envVal),
+	}
+	if _, err := buildImageFromContextWithArgs(imgName, ctx, args...); err != nil {
+		c.Fatal(err)
+	}
+
+	imgNameCache := "bldenvtestcachemiss"
```

Comment:
Same here, just twiddle your 'args' and rerun buildimage.... and check the IDs.  Much less code.